src/MatrixFields/matrix_field_utils.jl

```diff
@@ -0,0 +1,141 @@
+               diagonals with indices in the range $matrix_ld:$matrix_ud")
+
+    return n_rows, n_cols, matrix_ld, matrix_ud
+end
+
+"""
+    column_field2array(field)
+
+Converts a field defined on a `FiniteDifferenceSpace` into a `Vector` or a
+`BandedMatrix`, depending on whether or not the elements of the field are
+`BandMatrixRow`s. This involves copying the data stored in the field. Because
+`BandedMatrix` does not currently support operations with `CuArray`s, all GPU
+data is copied to the CPU.
+"""
+function column_field2array(field::Fields.FiniteDifferenceField)
+    if eltype(field) <: BandMatrixRow # field represents a matrix
+        n_rows, n_cols, matrix_ld, matrix_ud = banded_matrix_info(field)
+        matrix = BandedMatrix{eltype(eltype(field))}(
+            undef,
+            (n_rows, n_cols),
+            (-matrix_ld, matrix_ud),
+        )
+        for (index_of_field_entry, matrix_d) in enumerate(matrix_ld:matrix_ud)
+            matrix_diagonal = view(matrix, band(matrix_d))
+            diagonal_field = field.entries.:($index_of_field_entry)
+            diagonal_data =
+                vec(reinterpret(eltype(eltype(field)), parent(diagonal_field)'))
+
+            # Find the rows for which diagonal_data[row] is in the matrix.
+            # Note: The matrix index (1, 1) corresponds to the diagonal index 0,
+            # and the matrix index (n_rows, n_cols) corresponds to the diagonal
+            # index n_cols - n_rows.
+            first_row = matrix_d < 0 ? 1 - matrix_d : 1
+            last_row = matrix_d < n_cols - n_rows ? n_rows : n_cols - matrix_d
+
+            diagonal_data_view = view(diagonal_data, first_row:last_row)
+            @allowscalar copyto!(matrix_diagonal, diagonal_data_view)
+        end
+        return matrix
+    else # field represents a vector
+        return @allowscalar Array(column_field2array_view(field))
+    end
+end
+
+"""
+    column_field2array_view(field)
+
+Similar to `column_field2array(field)`, except that this version avoids copying
+the data stored in the field.
+"""
+function column_field2array_view(field::Fields.FiniteDifferenceField)
+    if eltype(field) <: BandMatrixRow # field represents a matrix
+        _, n_cols, matrix_ld, matrix_ud = banded_matrix_info(field)
+        field_data_transpose =
+            reinterpret(eltype(eltype(field)), parent(field)')
+        matrix_transpose =
+            _BandedMatrix(field_data_transpose, n_cols, matrix_ud, -matrix_ld)
+        return permutedims(matrix_transpose)
+        # TODO: Despite not copying any data, this function still allocates a
+        # small amount of memory because of _BandedMatrix and permutedims.
+    else # field represents a vector
+        return vec(reinterpret(eltype(field), parent(field)'))
+    end
+end
+
+all_columns(::Fields.FiniteDifferenceField) = (((1, 1), 1),)
+all_columns(field::Fields.ExtrudedFiniteDifferenceField) =
+    (Spaces.all_nodes(Spaces.horizontal_space(axes(field)))...,)
```

Comment:
Does this need to be splatted? Splatting large collections can be very costly. `all_nodes` returns an iterator, can we not map over it?

Reply:
I had this return a tuple to avoid allocations in the `matrix_field_broadcasting.jl` reference benchmarks. However, it looks like I can also avoid those allocations by using a `for`-loop over the arrays instead of calling `foreach`, so I'll do that instead and remove the splatting.